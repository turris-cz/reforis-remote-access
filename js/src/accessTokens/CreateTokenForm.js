/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    TextInput,
    Button,
    useAPIPost,
    useForm,
    useAlert,
    undefinedIfEmpty,
    API_STATE,
    formFieldsSize,
} from "foris";
import PropTypes from "prop-types";

import API_URLs from "API";

CreateTokenForm.propTypes = {
    generating: PropTypes.bool.isRequired,
    setGenerating: PropTypes.func.isRequired,
};

export default function CreateTokenForm({ generating, setGenerating }) {
    const [setAlert] = useAlert();

    const [createTokenResponse, createToken] = useAPIPost(API_URLs.tokens);
    useEffect(() => {
        if (createTokenResponse.state === API_STATE.ERROR) {
            setAlert(createTokenResponse.data);
            setGenerating(false);
        }
    }, [createTokenResponse, setAlert, setGenerating]);

    const [formState, formChangeHandler, reloadForm] =
        useForm(createTokenValidator);
    const formData = formState.data;
    const formErrors = formState.errors || {};
    useEffect(() => {
        reloadForm({ name: "" });
    }, [reloadForm]);

    function handleSubmit(event) {
        event.preventDefault();
        createToken({ data: formData });
        setGenerating(true);
    }

    if (!formData) {
        return null;
    }

    const addButtonDisabled = undefinedIfEmpty(formErrors) || generating;
    return (
        <div className={formFieldsSize}>
            <h2>{_("Create Token")}</h2>
            <form onSubmit={handleSubmit}>
                <TextInput
                    label={_("Token name")}
                    helpText={_(
                        `Shorter than 64 characters. Only alphanumeric characters, dots, dashes and underscores.`
                    )}
                    value={formData.name}
                    error={formErrors.name}
                    onChange={formChangeHandler((value) => ({
                        name: { $set: value },
                    }))}
                />
                <div className="text-end">
                    <Button
                        data-testid="add-token-button"
                        type="submit"
                        forisFormSize
                        disabled={addButtonDisabled}
                    >
                        {_("Add")}
                    </Button>
                </div>
            </form>
        </div>
    );
}

function createTokenValidator(formData) {
    const { name } = formData;
    if (!name) {
        return { name: _("Name cannot be empty") };
    }
    if (name.length > 64) {
        return { name: _("Name is too long") };
    }
    if (!/^[a-zA-Z0-9.\-_]+$/.test(name)) {
        return { name: _("Name contains invalid characters") };
    }
    return undefined;
}
