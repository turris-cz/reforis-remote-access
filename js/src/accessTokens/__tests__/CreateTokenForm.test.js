/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText, getByLabelText, fireEvent, wait } from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import CreateTokenForm from "../CreateTokenForm";

function getFormElements(container) {
    return {
        nameInput: getByLabelText(container, "Token name"),
        submitButton: getByText(container, "Add"),
    };
}

describe("<CreateTokenForm />", () => {
    let container;
    const setGenerating = jest.fn();

    beforeEach(() => {
        ({ container } = render(<CreateTokenForm generating={false} setGenerating={setGenerating} />));
    });

    it("should be rendered", () => {
        expect(container).toMatchSnapshot();
    });

    it("should check if name is empty", () => {
        const { nameInput, submitButton } = getFormElements(container);
        fireEvent.change(nameInput, { target: { value: "" } });
        expect(getByText(container, "Name cannot be empty")).toBeDefined();
        expect(submitButton.disabled).toBe(true);
    });

    it("should check if name has proper length", () => {
        const { nameInput, submitButton } = getFormElements(container);
        fireEvent.change(nameInput, { target: { value: "q".repeat(65) } });
        expect(getByText(container, "Name is too long")).toBeDefined();
        expect(submitButton.disabled).toBe(true);
    });

    it("should check if name contains invalid characters", () => {
        const { nameInput, submitButton } = getFormElements(container);
        fireEvent.change(nameInput, { target: { value: "!@#$%" } });
        expect(getByText(container, "Name contains invalid characters")).toBeDefined();
        expect(submitButton.disabled).toBe(true);
    });

    it("should add client", () => {
        const { nameInput, submitButton } = getFormElements(container);

        expect(submitButton.disabled).toBe(true);
        fireEvent.change(nameInput, { target: { value: "propername" } });
        expect(nameInput.value).toBe("propername");
        expect(submitButton.disabled).toBe(false);

        fireEvent.click(submitButton);
        expect(setGenerating).toBeCalledWith(true);
        expect(mockAxios.post).toBeCalledWith(
            "/reforis/remote-access/api/tokens",
            { name: "propername" },
            expect.anything(),
        );
    });

    it("should handle API error", async () => {
        const { nameInput, submitButton } = getFormElements(container);

        fireEvent.change(nameInput, { target: { value: "propername" } });
        fireEvent.click(submitButton);

        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
        });
        expect(setGenerating).toBeCalledWith(false);
    });
});
