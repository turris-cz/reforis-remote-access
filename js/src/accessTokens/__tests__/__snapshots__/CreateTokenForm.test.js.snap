// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateTokenForm /> should be rendered 1`] = `
<div>
  <h3>
    Create new token
  </h3>
  <form>
    <div
      class="form-group"
    >
      <label
        for="1"
      >
        Token name
      </label>
      <div
        class="input-group"
      >
        <input
          class="form-control  is-invalid"
          id="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="invalid-feedback"
      >
        Name cannot be empty
      </div>
      <small
        class="form-text text-muted"
      >
        Shorter than 64 characters. Only alphanumeric characters, dots, dashes and underscores.
      </small>
    </div>
    <button
      class="btn btn-primary  col-sm-12 col-lg-3"
      disabled=""
      type="submit"
    >
       
       
      Add
    </button>
  </form>
</div>
`;
